### PR TITLE
State Controller Update Patch.

### PR DESCRIPTION
**Pull request by nishantshreshth**

### CODE REVIEW

**Comment by nicolas33** [comment]
- Commit titles must not have your name.
- Only the maintainers are allowed to make commits for merges. Merging is the job of the maintainers, not the contributors.
- Your repository was **NOT** up-to-date.
- Don't work on the mainlines.

File: poc-07.py [hunk 1]
```diff
@@ -65,13 +65,36 @@ def search(self):
 
     def update(self, messages):
         """Update the storage.
-
```
Comment (nicolas33) [inline poc-07.py:68]: No. This is breaking the style of the Python comments in this file.

File: poc-07.py [hunk 2]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
         messages: is the collection of messages we have to create, update
         or remove."""
 
-        pass #TODO
-
-
+        """
```
Comment (nicolas33) [inline poc-07.py:71]: Wrong comments.
- """ are for describing Objects, API. They are exported by sphinx.
- # are for comments in the code.

File: poc-07.py [hunk 3]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+        Bug: If we provide try to update leftside first it will be updated as per right side changes provided by Engine. 
+             For Example: m2l is marked read and m2r is unread. Now when we are first trying to update left side 
+             with changes. We are making it unread on both the side.
+             Where as if we first update right side we will have m2l as well as m2r as read.
+             same goes for removal. Lets say there is m4l but m4r is missing . So what should we do should we remove m4l
+             or what.
```
Comment (nicolas33) [inline poc-07.py:77]: Too long lines. Set tw=80 in vim.

File: poc-07.py [hunk 4]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+        """
+        for message in messages:
+            if message not in self.messages:
+                self.messages.append(message)  #Adding new message to the sel.messages storage
```
Comment (nicolas33) [inline poc-07.py:81]: - Missing space after #.
- Missing a dot at the end.

File: poc-07.py [hunk 5]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+            else:
```
Comment (nicolas33) [inline poc-07.py:82]: Extra whitespace at the end of the line.

File: poc-07.py [hunk 6]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+             with changes. We are making it unread on both the side.
+             Where as if we first update right side we will have m2l as well as m2r as read.
+             same goes for removal. Lets say there is m4l but m4r is missing . So what should we do should we remove m4l
+             or what.
+        """
+        for message in messages:
+            if message not in self.messages:
+                self.messages.append(message)  #Adding new message to the sel.messages storage
+            else: 
+                for s_message  in self.messages:
```
Comment (nicolas33) [inline poc-07.py:83]: - Extra space before 'in'.
- Extra spaces the end of the line.
- "s_message": poor variable name.

File: poc-07.py [hunk 7]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+                self.messages.append(message)  #Adding new message to the sel.messages storage
+            else: 
+                for s_message  in self.messages:            
+                    if message.uid == s_message.uid and message.identical(s_message):
+                        break   
+                        """
+                        Message found in self.messages and is identical. 
+                        No need to search for message in self.messages and hence get out of loop
+                        """
+                    elif message.uid == s_message.uid and not message.identical(s_message):
```
Comment (nicolas33) [inline poc-07.py:90]: Poor conditional construct. Should be:

```
if a:
    if b:
        # code
    else:
        # code
```

File: poc-07.py [hunk 8]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+            if message not in self.messages:
+                self.messages.append(message)  #Adding new message to the sel.messages storage
+            else: 
+                for s_message  in self.messages:            
+                    if message.uid == s_message.uid and message.identical(s_message):
+                        break   
+                        """
+                        Message found in self.messages and is identical. 
+                        No need to search for message in self.messages and hence get out of loop
+                        """
```
Comment (nicolas33) [inline poc-07.py:89]: - Wrong comment: """ -> #

File: poc-07.py [hunk 9]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+                    elif message.uid == s_message.uid and not message.identical(s_message):
+                        """
+                        Updating the message.
+                        """
```
Comment (nicolas33) [inline poc-07.py:93]: - Ditto: wrong comment.

File: poc-07.py [hunk 10]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+                        s_message.body = message.body
+                        s_message.flags = message.flags
+                        break     
+
```
Comment (nicolas33) [inline poc-07.py:97]: Extra spaces for both above lines.

File: poc-07.py [hunk 11]
```diff
@@ -79,6 +102,13 @@ class StateBackend(Driver):
     def __init__(self):
         self.messages = Messages() # Fake the real data.
 
+    def update(self,mess):
```
Comment (nicolas33) [inline poc-07.py:105]: - Missing space after ','.
- Poor variable name.

File: poc-07.py [hunk 12]
```diff
@@ -65,13 +65,36 @@ def search(self):
...
+             Where as if we first update right side we will have m2l as well as m2r as read.
+             same goes for removal. Lets say there is m4l but m4r is missing . So what should we do should we remove m4l
+             or what.
+        """
+        for message in messages:
+            if message not in self.messages:
+                self.messages.append(message)  #Adding new message to the sel.messages storage
+            else: 
+                for s_message  in self.messages:            
+                    if message.uid == s_message.uid and message.identical(s_message):
```
Comment (Deepanshu2017) [inline poc-07.py:84]: It is guaranteed that the messages provided to driver.update() cannot be identical to any message present in driver (if they are then there is something wrong with fetch mechanism of getting changes) so entire if-else block can be converted into one if block.

File: poc-07.py [hunk 13]
```diff
@@ -79,6 +102,13 @@ class StateBackend(Driver):
     def __init__(self):
         self.messages = Messages() # Fake the real data.
 
+    def update(self,mess):
+        self.messages = mess
+
+    """
+        StateBackend Must Contain MetaData for last synced messages rather than messages.
+        For now we are putting messages. WE HAVE TO LATER THINK OF ITS IMPLEMENTATION.
+    """
```
Comment (nicolas33) [inline poc-07.py:111]: Correct comment but at the wrong place and with bad indentation.

File: poc-07.py [hunk 14]
```diff
@@ -91,15 +121,18 @@ class StateController(object):
         - the engine;
         - our state backend;
         - their state backend (read-only).
+        #Comment Nishant : I think State Controller is supposed to communicate with thier state backend 
+        (write - Only)
```
Comment (nicolas33) [inline poc-07.py:125]: Why not just fix it?

File: poc-07.py [hunk 15]
```diff
@@ -91,15 +121,18 @@ class StateController(object):
...
     """
 
     def __init__(self, driver):
         self.driver = driver # The driver we own.
         self.state = StateBackend() # Would be an emitter.
 
-    def update(self, theirMessages):
+    def update(self, theirMessages, theirstate):
```
Comment (nicolas33) [inline poc-07.py:132]: - variable name does not respect the style of the code base (camelCase).
- Is it the good time to get "theirState"?

File: poc-07.py [hunk 16]
```diff
@@ -91,15 +121,18 @@ class StateController(object):
...
         """Update this side with the messages from the other side."""
-        pass #TODO: compare their messages with what we have in this repository.
+        self.driver.update(theirMessages)
+        theirstate.update(self.driver.messages)
```
Comment (nicolas33) [inline poc-07.py:135]: What if one message can't be updated by the driver (exception raised)?

File: poc-07.py [hunk 17]
```diff
@@ -117,6 +150,7 @@ def search(self):
                     if message.uid == stateMessage.uid:
                         if not message.identical(stateMessage):
                             changedMessages.append(message)
+                            break #There is no point of iterating further.
```
Comment (nicolas33) [inline poc-07.py:153]: Good point. Should have a space after '#".

File: poc-07.py [hunk 18]
```diff
@@ -150,8 +184,8 @@ def run(self):
         print("left: %s"% leftMessages)
         print("rght: %s"% rightMessages)
 
-        self.left.update(rightMessages)
-        self.right.update(leftMessages)
+        self.left.update(rightMessages,self.right.state)
+        self.right.update(leftMessages,self.left.state)
```
Comment (nicolas33) [inline poc-07.py:188]: Missing a space after ',' for two above lines.

**Pull request merged**